subscriber_notifee.go

```diff
@@ -0,0 +1,126 @@
+package dht
+
+import (
+	"github.com/jbenet/goprocess"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	"github.com/libp2p/go-eventbus"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// subscriberNotifee implements network.Notifee and also manages the subscriber to the event bus. We consume peer
+// identification events to trigger inclusion in the routing table, and we consume Disconnected events to eject peers
+// from it.
+type subscriberNotifee IpfsDHT
+
+func (nn *subscriberNotifee) DHT() *IpfsDHT {
+	return (*IpfsDHT)(nn)
+}
+
+func (nn *subscriberNotifee) subscribe(proc goprocess.Process) {
+	dht := nn.DHT()
+
+	dht.host.Network().Notify(nn)
+	defer dht.host.Network().StopNotify(nn)
+
+	var err error
+	evts := []interface{}{
+		&event.EvtPeerIdentificationCompleted{},
+	}
+
+	// subscribe to the EvtPeerIdentificationCompleted event which notifies us every time a peer successfully completes identification
+	sub, err := dht.host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+	if err != nil {
+		logger.Errorf("dht not subscribed to peer identification events; things will fail; err: %s", err)
+	}
+	defer sub.Close()
+
+	for {
+		select {
+		case evt, more := <-sub.Out():
+			// we will not be getting any more events
+			if !more {
+				return
+			}
+
+			// something has gone really wrong if we get an event for another type
+			ev, ok := evt.(event.EvtPeerIdentificationCompleted)
+			if !ok {
+				logger.Errorf("got wrong type from subscription: %T", ev)
+				return
+			}
+
+			// if the peer supports the DHT protocol, add it to our RT and kick a refresh if needed
+			protos, err := dht.peerstore.SupportsProtocols(ev.Peer, dht.protocolStrs()...)
+			if err == nil && len(protos) != 0 {
+				refresh := dht.routingTable.Size() <= minRTRefreshThreshold
+				dht.Update(dht.ctx, ev.Peer)
+				if refresh && dht.autoRefresh {
+					select {
+					case dht.triggerRtRefresh <- nil:
+					default:
+					}
+				}
+			}
+
+		case <-proc.Closing():
+			return
+		}
+	}
+}
+
+func (nn *subscriberNotifee) Disconnected(n network.Network, v network.Conn) {
+	dht := nn.DHT()
+	select {
+	case <-dht.Process().Closing():
+		return
+	default:
+	}
+
+	p := v.RemotePeer()
+
+	// Lock and check to see if we're still connected. We lock to make sure
+	// we don't concurrently process a connect event.
+	dht.plk.Lock()
```

Comment:
@Stebalien Shouldn't we also take this lock in the `IdentifyCompleted` loop above ? Why are we synchronising ONLY across Disconnects here ?

We _could_ end up processing an IdentifyCompleted event for a Conn2 between the peers BEFORE we process a Disconnect for a prior Conn1.

Reply:
Maybe another question do we need this lock at all?

If I understand correctly we used to need this lock because we weren't using `peerstore.SupportsProtocol` to check if we should add/remove them from the routing table. As long as `peerstore.SupportsProtocol` returns the correct info it seems like we should be fine.

Reply:
Hm. @aarshkshah1992 is right. We need to take this lock _and_ we need to check to see if we're connected (under the lock) before updating the routing table. Otherwise, we could process the disconnect event _before_ the identify event, and add a dead peer to the routing table.

On the other hand, this issue should go away with the new routing table changes, right?

Reply:
There are two cases:

1. If a live peer gets marked as "missing" in the RT, it will get fixed after 2 minutes with a connectdness check. This shouldn't be a big deal.
2. If a dead peer gets marked as live in the RT, it will remain there & squat capacity till it's explicitly marked as dead by the caller ( currently, that would happen when we try to send a message to it and get a `msmux.ErrNotSupported`). This _can_ create problems . 

We should lock.

Reply:
Think this was handled in the latest update